sandbox/grist/imports/import_csv.py

```diff
@@ -90,108 +90,117 @@ def parse_file(file_path, parse_options=None):
-      headers = [''] * len(headers)
-
-    row_set.register_processor(messytables.offset_processor(data_offset))
-    rows = [
-      [cell.value for cell in row]
-      for row in row_set
-    ]
-    table_data_with_types = parse_data.get_table_data(rows, len(headers), num_rows)
-
-    # Identify and remove empty columns, and populate separate metadata and data lists.
-    column_metadata = []
-    table_data = []
-    for col_data, header in zip(table_data_with_types, headers):
-      if not header and all(val == "" for val in col_data["data"]):
-        continue # empty column
-      data = col_data.pop("data")
-      col_data["id"] = header
-      column_metadata.append(col_data)
-      table_data.append(data)
-
-    if not table_data:
-      # Don't add tables with no columns.
-      continue
-
-    guessed = row_set._dialect
-    quoting = parse_options.get('quoting')
-    options = {"delimiter": parse_options.get('delimiter', guessed.delimiter),
-               "doublequote": parse_options.get('doublequote', guessed.doublequote),
-               "lineterminator": parse_options.get('lineterminator', guessed.lineterminator),
-               "quotechar": parse_options.get('quotechar', guessed.quotechar),
-               "skipinitialspace": parse_options.get('skipinitialspace', guessed.skipinitialspace),
-               "include_col_names_as_headers": include_col_names_as_headers,
-               "start_with_row": 1,
-               "NUM_ROWS": num_rows,
-               "SCHEMA": SCHEMA
-               }
-
-    log.info("Output table %r with %d columns", table_name, len(column_metadata))
-    for c in column_metadata:
-      log.debug("Output column %s", c)
-    export_list.append({
-      "table_name": table_name,
-      "column_metadata": column_metadata,
-      "table_data": table_data
-    })
+
+  table_name = None
+  rows = list(reader)
+  sample_len = 100
+  sample_rows = rows[:sample_len]
+  data_offset, headers = import_utils.headers_guess(sample_rows)
+
+  # Make sure all header values are strings.
+  for i, header in enumerate(headers):
+    if not isinstance(header, six.string_types):
+      headers[i] = six.text_type(header)
+
+  log.info("Guessed data_offset as %s", data_offset)
+  log.info("Guessed headers as: %s", headers)
+
+  have_guessed_headers = any(headers)
+  include_col_names_as_headers = parse_options.get('include_col_names_as_headers',
+                                                   have_guessed_headers)
+
+  if include_col_names_as_headers and not have_guessed_headers:
+    # use first line as headers
+    data_offset, first_row = import_utils.find_first_non_empty_row(sample_rows)
+    headers = import_utils.expand_headers(first_row, data_offset, sample_rows)
+
+  elif not include_col_names_as_headers and have_guessed_headers:
+    # move guessed headers to data
+    data_offset -= 1
+    headers = [''] * len(headers)
+
+  rows = rows[data_offset:]  # Use row.pop instead to make it faster ?
```

Comment:
I think this method is fine.

Reply:
```suggestion
  rows = rows[data_offset:]
```

Reply:
Done!
Github does not let me apply suggestion today, don't know why exactly:

![image](https://user-images.githubusercontent.com/146023/191089921-0b21806a-4efd-42b5-90ac-171a5120f2a6.png)